### PR TITLE
Constrain dependencies to our requirements.txt file

### PR DESCRIPTION
**Pull request by A5rocks**
Fixes https://github.com/python-trio/trio/issues/3057

### CODE REVIEW

**Comment by A5rocks** [comment]
Comparing https://github.com/python-trio/trio/actions/runs/10294036703/job/28491403489 (latest run on main branch) to https://github.com/python-trio/trio/actions/runs/10294184496/job/28491802374?pr=3061 (latest run on this branch):

main branch:
```
  Successfully installed attrs-24.2.0 exceptiongroup-1.2.2 idna-3.7 outcome-1.3.0.post0 sniffio-1.3.1 sortedcontainers-2.4.0 trio-0.26.2+dev
```

this branch:
```
  Successfully installed attrs-24.1.0 exceptiongroup-1.2.1 idna-3.7 outcome-1.3.0.post0 sniffio-1.3.1 sortedcontainers-2.4.0 trio-0.26.2+dev
```

Of importance, this branch uses `attrs-24.1.0` rather than the newly released `attrs-24.2.0` (which removes the deprecation warning), as well as an older exceptiongroup, etc. This means that the no-test-deps CI job won't start failing in the middle of nowhere, instead it will only fail for autodeps runs.

File: ci.sh [hunk 1]
```diff
@@ -52,7 +52,7 @@ else
     # expands to 0 != 1 if NO_TEST_REQUIREMENTS is not set, if set the `-0` has no effect
     # https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02
     if [ ${NO_TEST_REQUIREMENTS-0} == 1 ]; then
-        python -m pip install pytest coverage
+        python -m pip install pytest coverage -c test-requirements.txt
         flags="--skip-optional-imports"
     else
         python -m pip install -r test-requirements.txt
```
Comment (CoolCat467) [inline ci.sh:58]: I think this is what you intended on doing:
```suggestion
        python -m pip install -c test-requirements.txt
```

**Review by CoolCat467 — CHANGES_REQUESTED** [review]

**Comment by A5rocks** [comment]
-c just means giving a file as constraints, not installing it as a requirements file

**Review by CoolCat467 — APPROVED** [review]
With information I have learned, I think this looks pretty good

**Pull request merged**